src/expr.rs

```diff
@@ -64,7 +64,9 @@ pub(crate) fn format_expr(
     if contains_skip(&*expr.attrs) {
         return Some(context.snippet(expr.span()).to_owned());
     }
-    let shape = if expr_type == ExprType::Statement && semicolon_for_expr(context, expr) {
+    let shape = if expr_type == ExprType::Statement
+        && semicolon_for_expr_extra_hacky_double_counted_spacing(context, expr)
+    {
```

Comment:
I cannot remember what the bug policy here is, but ideally we'd just remove this adjustment to `shape` since it's already accounted for in `format_stmt`. That causes self-formatting test to fail because some expressions actually fit onto one line.

Reply:
I fixed it in Version=Two, added a test for one&two, and fixed the cases that changed rustfmt's own formatting (since rustfmt apparently uses Two! 😆)

Reply:
responded to this in https://github.com/rust-lang/rustfmt/pull/6149#discussion_r1580147537